Tendermint Light Client Documentation (#6801)
cosmos/ibc-go: docs/docs/03-light-clients/05-tendermint/01-overview.md

Reviewer: Are there more files coming? If not maybe this should be broken down to different files?

Author: Not from myself. but future tendermint docs can come in and be in new files

---
title: Overview
sidebar_label: Overview
sidebar_position: 1
slug: /ibc/light-clients/tendermint/overview
---

# `07-tendermint`

## Overview

> The tendermint client is the first and most deployed light client in IBC. It implements the IBC Light Client interface to track a counterparty running Comet BFT consensus. Note: Tendermint is the old name of Comet BFT which has been retained in IBC to avoid expensive migration costs.

Reviewer 2: no harm to throw in a link here to the cometbft repo

```suggestion
The tendermint client is the first and most deployed light client in IBC. It implements the IBC Light Client interface to track a counterparty running [CometBFT](https://github.com/cometbft/cometbft) consensus. Note: Tendermint is the old name of CometBFT which has been retained in IBC to avoid expensive migration costs.
```

sidenote, I think they still refer to cometbft as a implementation running "Tendermint consensus", it might be good to clarify that before we merge this

> The tendermint client is the first and most deployed light client in IBC. It implements the IBC Light Client interface to track a counterparty running Comet BFT consensus. Note: Tendermint is the old name of Comet BFT which has been retained in IBC to avoid expensive migration costs.

Reviewer: I think it is important to give some technical context in the overview. For example, the features it supports and how it differs from a basic header tracking light client. Mentioning support of historical proofs and misbehaviors.

Possibly mention the state the needs to be stored by the light client i.e.:

- Client State (as parameters)
- A mapping from Height -> Consensus State

Otherwise it can be difficult for the first time reader to understand client state is stored as a singleton but consensus states are stored as maps later on.

## Synopsis

### Initialization

> The Tendermint light client is initialized with a ClientState that contains parameters necessary for CometBFT header verification along with a latest height and ConsensusState which encapsulates a root of trust header that will serve to verify future incoming headers from the counterparty.

Reviewer 2: ```suggestion
The Tendermint light client is initialized with a ClientState that contains parameters necessary for CometBFT header verification along with a latest height and ConsensusState which encapsulates the application state root of a trust header that will serve to verify future incoming headers from the counterparty.
```

Author: Sorry I don't mean application root here, i mean the initial header itself serves as a root-of-trust. Will hyphenate for clarity

  // Latest height the client was updated to
  ibc.core.client.v1.Height latest_height = 7;

  // Proof specifications used in verifying counterparty state
  repeated cosmos.ics23.v1.ProofSpec proof_specs = 8;

  // Path at which next upgraded client will be committed.
  // Each element corresponds to the key for a single CommitmentProof in the
  // chained proof. NOTE: ClientState must stored under
  // `{upgradePath}/{upgradeHeight}/clientState` ConsensusState must be stored
  // under `{upgradepath}/{upgradeHeight}/consensusState` For SDK chains using
  // the default upgrade module, upgrade_path should be []string{"upgrade",
  // "upgradedIBCState"}`
  repeated string upgrade_path = 9;
}
```

```proto
message ConsensusState {
  // timestamp that corresponds to the block height in which the ConsensusState
  // was stored.
  google.protobuf.Timestamp timestamp = 1;
  // commitment root (i.e app hash) that will be used
  // to verify proofs of packet flow messages
  ibc.core.commitment.v1.MerkleRoot root = 2;
  // hash of the next validator set that will be used as
  // a new updated source of trust to verify future updates
  bytes next_validators_hash = 3;
}
```

### Updates

> Once the initial client state and consensus state are submitted, future consensus states can be added to the client by submitting IBC Headers. These headers contain all necessary information to run the Comet BFT Light Client protocol. For detailed information on the Comet BFT Light Client protocol and its safety properties please refer to the Comet BFT paper.

Reviewer 2: should we add a link to the paper?

```proto
message Header {

>   // this is the new signed header that we want to add
>   // as a consensus state to the client

Reviewer 2: ```suggestion
  // this is the new signed header that we want to add
  // as a new consensus state to the ibc client. 
  // the signed header contains the commit signatures of the `validator_set` below
```

  .tendermint.types.SignedHeader signed_header = 1;

>   // this is the 

Reviewer 2: ```suggestion
  // the validator set which signed the new header
```

  .tendermint.types.ValidatorSet validator_set      = 2;

>   ibc.core.client.v1.Height      trusted_height     = 3;

Reviewer 2: ```suggestion
  // the trusted height of the consensus state which we are updating from
  ibc.core.client.v1.Height      trusted_height     = 3;
```

>   .tendermint.types.ValidatorSet trusted_validators = 4;

Reviewer 2: ```suggestion
  // the trusted validator set, the hash of the trusted validators must be equal to 
  // `next_validators_hash` of the current consensus state
  .tendermint.types.ValidatorSet trusted_validators = 4;
```
